# Drought_pCRE

Abstract


Water deficit is a significant limiting factor to agricultural production. Plants respond to the drought stress by multiple layers of changes including morphology, physiology and molecular. Modulating gene expression is an effective way for plants to cope with stress and recover. Many transcription factors (TF), especially ABA- dependent, have been proved involved in drought stress response, yet more TFs relative to other physiological activities such as responses to light, temperature and oxygen are remained largely unknown.  The cis-regulatory elements (CREs), known as TF binding sites, and their combinations provide a powerful approach to decode unknown TFs and their regulatory network. Here we profiled the transcriptome changes by RNA-seq in plants stressed by drought and recovered for a series of time. All differentially regulated genes were grouped into 40 regulatory patterns, and more than 2,000 pCREs in the promoters of these genes were discovered. We then trained machine learning models to predict their response by using pCREs as features. The top10 most important pCREs in each gene cluster were matched with TF binding sites database to find corresponding TFs. This study demonstrated how cis-regulatory information as well as machine learning can help finding new TFs regulating plant response to drought stress.   




References:


Azodi, Christina B., John P. Lloyd, and Shin-Han Shiu. "The cis-regulatory codes of response to combined heat and drought stress in Arabidopsis thaliana." NAR genomics and bioinformatics 2.3 (2020): lqaa049.
Moore, Bethany M., et al. "Modeling gene regulation in response to wounding: temporal variations, hormonal variations, and specialized metabolism pathways induced by wounding." bioRxiv (2020).
Schwarz, Birte, et al. "Putative cis-regulatory elements predict iron deficiency responses in Arabidopsis roots." Plant physiology 182.3 (2020): 1420-1439.
